api/envoy/extensions/filters/http/custom_response/v3/custom_response.proto

```diff
@@ -0,0 +1,76 @@
+syntax = "proto3";
+
+package envoy.extensions.filters.http.custom_response.v3;
+
+import "envoy/config/core/v3/base.proto";
+import "envoy/config/core/v3/http_uri.proto";
+import "envoy/config/core/v3/substitution_format_string.proto";
+
+import "google/protobuf/wrappers.proto";
+
+import "xds/annotations/v3/status.proto";
+import "xds/type/matcher/v3/matcher.proto";
+
+import "udpa/annotations/status.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.filters.http.custom_response.v3";
+option java_outer_classname = "CustomResponseProto";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/custom_response/v3;custom_responsev3";
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+option (xds.annotations.v3.file_status).work_in_progress = true;
+
+// [#protodoc-title: Custom Response Filter]
+// Custom Response Filter
+// [#extension: envoy.filters.http.custom_response]
+
+// The Custom Response Filter allows for replacing upstream responses.
+
+// The filter configuration is a collection of custom response
+// policies.
+message CustomResponse {
+  // Definition of the custom response to be provided to the downstream.
+  // [#next-free-field: 7]
+  message Response {
+    message RemoteDataSource {
+      config.core.v3.HttpUri http_uri = 1;
+
+      config.core.v3.RetryPolicy retry_policy = 2;
+    }
```

Comment:
Seems it would be better if we make this a downgrade mechanism. If the upstream reply match specified pattern, then we can send the original request to another remote service or just return a local response.

So, IMO, we should get the whole response from the `remote`, not just a body.

Reply:
Hi @wbpcode, I have modified the implementation in a separate PR (#23084) that does this.